diff --git a/camera/device/3.2/ICameraDeviceSession.hal b/camera/device/3.2/ICameraDeviceSession.hal
index 477a3cc5..118042ed 100644
--- a/camera/device/3.2/ICameraDeviceSession.hal
+++ b/camera/device/3.2/ICameraDeviceSession.hal
@@ -27,6 +27,30 @@ import android.hardware.camera.common@1.0::types;
  */
 interface ICameraDeviceSession {
 
+    /**
+     * getDeviceVersion:
+     *
+     * To support 3.0 device, expose version by this
+     *
+     * Return values:
+     * @return version The HAL device version
+     *
+     */
+    getDeviceVersion() generates (uint32_t version);
+
+    /**
+     * registerStreamBuffers:
+     *
+     * To support 3.0 device, port function from Android N
+     * Camera3 device may not use the bufferSet again, no need to sendback
+     *
+     * Return values:
+     * @return status Status code for the operation
+     *
+     */
+    registerStreamBuffers(StreamBufferSet hidlBufferSet) generates
+            (Status status);
+
     /**
      * constructDefaultRequestSettings:
      *
diff --git a/camera/device/3.2/default/CameraDevice.cpp b/camera/device/3.2/default/CameraDevice.cpp
index 295ee328..16738f68 100644
--- a/camera/device/3.2/default/CameraDevice.cpp
+++ b/camera/device/3.2/default/CameraDevice.cpp
@@ -47,9 +47,15 @@ CameraDevice::CameraDevice(
 
     mDeviceVersion = mModule->getDeviceVersion(mCameraIdInt);
     if (mDeviceVersion < CAMERA_DEVICE_API_VERSION_3_2) {
-        ALOGE("%s: Camera id %s does not support HAL3.2+",
-                __FUNCTION__, mCameraId.c_str());
-        mInitFail = true;
+        //hack to force support 3.0
+        if (mDeviceVersion >= CAMERA_DEVICE_API_VERSION_3_0) {
+            ALOGW("%s: Camera id %s has version %x but HAL3.2+ is required on Android 8.0+, need forward-port",
+                    __FUNCTION__, mCameraId.c_str(), mDeviceVersion);
+        } else {
+            ALOGE("%s: Camera id %s does not support HAL3.0+",
+                    __FUNCTION__, mCameraId.c_str());
+            mInitFail = true;
+        }
     }
 }
 
@@ -218,6 +224,11 @@ Return<void> CameraDevice::open(const sp<ICameraDeviceCallback>& callback, open_
 
         /** Cross-check device version */
         if (device->common.version < CAMERA_DEVICE_API_VERSION_3_2) {
+            //hack to support 3.0
+            if (device->common.version >= CAMERA_DEVICE_API_VERSION_3_0) {
+            ALOGW("%s: Open camera version %x by camera_device %x, need forward-port",
+                    __FUNCTION__, device->common.version, CAMERA_DEVICE_API_VERSION_3_2);
+            } else {
             ALOGE("%s: Could not open camera: "
                     "Camera device should be at least %x, reports %x instead",
                     __FUNCTION__,
@@ -227,6 +238,7 @@ Return<void> CameraDevice::open(const sp<ICameraDeviceCallback>& callback, open_
             mLock.unlock();
             _hidl_cb(Status::ILLEGAL_ARGUMENT, nullptr);
             return Void();
+            }
         }
 
         struct camera_info info;
diff --git a/camera/device/3.2/default/CameraDeviceSession.cpp b/camera/device/3.2/default/CameraDeviceSession.cpp
index d6a04bc5..ed55c436 100644
--- a/camera/device/3.2/default/CameraDeviceSession.cpp
+++ b/camera/device/3.2/default/CameraDeviceSession.cpp
@@ -737,6 +737,49 @@ void CameraDeviceSession::ResultBatcher::processCaptureResult(CaptureResult& res
 }
 
 // Methods from ::android::hardware::camera::device::V3_2::ICameraDeviceSession follow.
+Return<uint32_t> CameraDeviceSession::getDeviceVersion() {
+    return mDeviceVersion;
+}
+
+Return<Status> CameraDeviceSession::registerStreamBuffers(
+        const StreamBufferSet& hidlBufferSet) {
+    Status status = initStatus();
+
+    //rebuild bufferSet
+    //from StreamBufferSet
+    camera3_stream_buffer_set_t bufferSet;
+    Camera3Stream stream;
+    convertFromHidl(hidlBufferSet.stream, &stream);
+
+    uint32_t numBufs = hidlBufferSet.numBuffers;
+
+    //convert the buffers of stream buffer
+    hidl_vec<buffer_handle_t*> allBufPtrs;
+    allBufPtrs.resize(numBufs);
+    for (size_t i = 0; i < numBufs; i++) {
+        buffer_handle_t buf = hidlBufferSet.buffers[i].getNativeHandle();
+        allBufPtrs[i] = &buf;
+    }
+
+    bufferSet.stream = &stream;
+    bufferSet.num_buffers = numBufs;
+    bufferSet.buffers = allBufPtrs.data();
+
+    ATRACE_BEGIN("camera3->register_stream_buffers");
+    status_t ret = mDevice->ops->register_stream_buffers(mDevice, &bufferSet);
+    ATRACE_END();
+
+    if (ret == OK) {
+        status = Status::OK;
+    }else if (ret == -EINVAL) {
+        status = Status::ILLEGAL_ARGUMENT;
+    } else if (ret != OK) {
+        status = Status::INTERNAL_ERROR;
+    }
+
+    return status;
+}
+
 Return<void> CameraDeviceSession::constructDefaultRequestSettings(
         RequestTemplate type, ICameraDeviceSession::constructDefaultRequestSettings_cb _hidl_cb)  {
     Status status = initStatus();
diff --git a/camera/device/3.2/default/CameraDeviceSession.h b/camera/device/3.2/default/CameraDeviceSession.h
index 69e2e2c8..a0cad37a 100644
--- a/camera/device/3.2/default/CameraDeviceSession.h
+++ b/camera/device/3.2/default/CameraDeviceSession.h
@@ -43,6 +43,7 @@ using ::android::hardware::camera::device::V3_2::CaptureRequest;
 using ::android::hardware::camera::device::V3_2::HalStreamConfiguration;
 using ::android::hardware::camera::device::V3_2::StreamConfiguration;
 using ::android::hardware::camera::device::V3_2::ICameraDeviceSession;
+using ::android::hardware::camera::device::V3_2::StreamBufferSet;
 using ::android::hardware::camera::common::V1_0::Status;
 using ::android::hardware::camera::common::V1_0::helper::HandleImporter;
 using ::android::hardware::kSynchronizedReadWrite;
@@ -95,6 +96,9 @@ protected:
 
     // Methods from ::android::hardware::camera::device::V3_2::ICameraDeviceSession follow
 
+    Return<uint32_t> getDeviceVersion();
+    Return<Status> registerStreamBuffers(
+            const StreamBufferSet& hidlBufferSet);
     Return<void> constructDefaultRequestSettings(
             RequestTemplate type,
             ICameraDeviceSession::constructDefaultRequestSettings_cb _hidl_cb);
@@ -323,6 +327,15 @@ private:
         TrampolineSessionInterface_3_2(sp<CameraDeviceSession> parent) :
                 mParent(parent) {}
 
+        virtual Return<uint32_t> getDeviceVersion() override {
+            return mParent->getDeviceVersion();
+        }
+
+        virtual Return<Status> registerStreamBuffers(
+                const V3_2::StreamBufferSet& hidlBufferSet) override {
+            return mParent->registerStreamBuffers(hidlBufferSet);
+        }
+
         virtual Return<void> constructDefaultRequestSettings(
                 V3_2::RequestTemplate type,
                 V3_2::ICameraDeviceSession::constructDefaultRequestSettings_cb _hidl_cb) override {
diff --git a/camera/device/3.2/types.hal b/camera/device/3.2/types.hal
index 276e92a3..8a00621e 100644
--- a/camera/device/3.2/types.hal
+++ b/camera/device/3.2/types.hal
@@ -509,6 +509,29 @@ struct StreamBuffer {
 
 };
 
+/**
+ * StreamBufferSet:
+ *
+ * This is to support 3.0 interface
+ *
+ */
+struct StreamBufferSet {
+    /**
+     * A Camera3Stream object
+     */
+    Stream stream;
+
+    /**
+     * The number of buffers
+     */
+    uint32_t numBuffers;
+
+    /**
+     * The the vector of buffer_handle from the stream
+     */
+    vec<handle> buffers;
+};
+
 /**
  * CameraBlob:
  *
diff --git a/camera/device/3.3/default/CameraDeviceSession.h b/camera/device/3.3/default/CameraDeviceSession.h
index dd52b35b..eee8a9c7 100644
--- a/camera/device/3.3/default/CameraDeviceSession.h
+++ b/camera/device/3.3/default/CameraDeviceSession.h
@@ -82,6 +82,15 @@ private:
         TrampolineSessionInterface_3_3(sp<CameraDeviceSession> parent) :
                 mParent(parent) {}
 
+        virtual Return<uint32_t> getDeviceVersion() override {
+            return mParent->getDeviceVersion();
+        }
+
+        virtual Return<Status> registerStreamBuffers(
+                const V3_2::StreamBufferSet& hidlBufferSet) override {
+            return mParent->registerStreamBuffers(hidlBufferSet);
+        }
+
         virtual Return<void> constructDefaultRequestSettings(
                 V3_2::RequestTemplate type,
                 V3_3::ICameraDeviceSession::constructDefaultRequestSettings_cb _hidl_cb) override {
diff --git a/camera/provider/2.4/default/CameraProvider.cpp b/camera/provider/2.4/default/CameraProvider.cpp
index d50168a2..3fcf3b76 100644
--- a/camera/provider/2.4/default/CameraProvider.cpp
+++ b/camera/provider/2.4/default/CameraProvider.cpp
@@ -37,6 +37,8 @@ namespace {
 const char *kLegacyProviderName = "legacy/0";
 // "device@<version>/legacy/<id>"
 const std::regex kDeviceNameRE("device@([0-9]+\\.[0-9]+)/legacy/(.+)");
+const char *kHAL3_0 = "3.0";
+const char *kHAL3_1 = "3.1";
 const char *kHAL3_2 = "3.2";
 const char *kHAL3_3 = "3.3";
 const char *kHAL1_0 = "1.0";
@@ -147,6 +149,10 @@ int CameraProvider::getCameraDeviceVersion(const hidl_string& deviceName) {
         return CAMERA_DEVICE_API_VERSION_3_3;
     } else if (deviceVersion == kHAL3_2) {
         return CAMERA_DEVICE_API_VERSION_3_2;
+    } else if (deviceVersion == kHAL3_1) {
+        return CAMERA_DEVICE_API_VERSION_3_1;
+    } else if (deviceVersion == kHAL3_0) {
+        return CAMERA_DEVICE_API_VERSION_3_0;
     } else if (deviceVersion == kHAL1_0) {
         return CAMERA_DEVICE_API_VERSION_1_0;
     }
@@ -157,6 +163,8 @@ std::string CameraProvider::getHidlDeviceName(
         std::string cameraId, int deviceVersion) {
     // Maybe consider create a version check method and SortedVec to speed up?
     if (deviceVersion != CAMERA_DEVICE_API_VERSION_1_0 &&
+            deviceVersion != CAMERA_DEVICE_API_VERSION_3_0 &&
+            deviceVersion != CAMERA_DEVICE_API_VERSION_3_1 &&
             deviceVersion != CAMERA_DEVICE_API_VERSION_3_2 &&
             deviceVersion != CAMERA_DEVICE_API_VERSION_3_3 &&
             deviceVersion != CAMERA_DEVICE_API_VERSION_3_4 ) {
@@ -214,6 +222,11 @@ bool CameraProvider::initialize() {
     mPreferredHal3MinorVersion = property_get_int32("ro.camera.wrapper.hal3TrebleMinorVersion", 3);
     ALOGV("Preferred HAL 3 minor version is %d", mPreferredHal3MinorVersion);
     switch(mPreferredHal3MinorVersion) {
+        case 0:
+        case 1:
+            ALOGW("Minor camera device HAL version %d in property "
+                    "'camera.wrapper.hal3TrebleMinorVersion' is for old HAL device", mPreferredHal3MinorVersion);
+            break;
         case 2:
         case 3:
             // OK
@@ -294,10 +307,13 @@ int CameraProvider::checkCameraVersion(int id, camera_info info) {
             case CAMERA_DEVICE_API_VERSION_3_4:
                 // in support
                 break;
-            case CAMERA_DEVICE_API_VERSION_2_0:
-            case CAMERA_DEVICE_API_VERSION_2_1:
             case CAMERA_DEVICE_API_VERSION_3_0:
             case CAMERA_DEVICE_API_VERSION_3_1:
+                ALOGW("%s: Device %d has HAL version %x, limited support by forward-port",
+                        __FUNCTION__, id, info.device_version);
+                break;
+            case CAMERA_DEVICE_API_VERSION_2_0:
+            case CAMERA_DEVICE_API_VERSION_2_1:
                 // no longer supported
             default:
                 ALOGE("%s: Device %d has HAL version %x, which is not supported",
@@ -485,6 +501,8 @@ Return<void> CameraProvider::getCameraDeviceInterface_V3_x(
     // system property.
     sp<android::hardware::camera::device::V3_2::ICameraDevice> device;
     switch (mPreferredHal3MinorVersion) {
+        case 0:
+        case 1:
         case 2: { // Map legacy camera device v3 HAL to Treble camera device HAL v3.2
             ALOGV("Constructing v3.2 camera device");
             sp<android::hardware::camera::device::V3_2::implementation::CameraDevice> deviceImpl =
diff --git a/current.txt b/current.txt
index db34c37b..cbaa1238 100644
--- a/current.txt
+++ b/current.txt
@@ -55,8 +55,8 @@ c1705e9d62438a1d955269965af915ae28e692bd480a3b1ce67056fef992d62f android.hardwar
 4db48439ce9dde97f1cfb3d7408f6c737f621ac0f7494aeea35ed599bc2352a3 android.hardware.camera.device@1.0::types
 b32f9aeaf1c442195eb06ffc7600968c919d005b2718874f09c57287fae55918 android.hardware.camera.device@3.2::ICameraDevice
 63bfc3da0f2d2301f7a0508c7c2b9ffc521d4d545ee03718da70e9d6273b3b21 android.hardware.camera.device@3.2::ICameraDeviceCallback
-0fa3e1e64819283b8737fc4e5ab759f0cb4ac1a996e8a51cc4aa8025a457208e android.hardware.camera.device@3.2::ICameraDeviceSession
-030be3d2b159cbde7920485807140f6b6064ef4a5de4a40a6c4bc8d2c72f7cd3 android.hardware.camera.device@3.2::types
+c2f5b5788f043755f1f71fcacda2aa64c00f4231d2ed6159e139429aa94f1cd7 android.hardware.camera.device@3.2::ICameraDeviceSession
+c7ff43776a552b693780987cfc7f6205e0cd2a2131ffb6ae0ec0d2a839637246 android.hardware.camera.device@3.2::types
 5ba7947cee515d7a2359bfcbfb9678c1c3a768c288471919ac095b96ae6f3d40 android.hardware.camera.metadata@3.2::types
 f7e299d85033ac52d1095a35784fcfeaff43603f58c751e4153c85bbade3b330 android.hardware.camera.provider@2.4::ICameraProvider
 a501ca1aecd09f1b9fd70a9af84205430dbd49a808e8fa395d363b9902e6f58c android.hardware.camera.provider@2.4::ICameraProviderCallback
