diff --git a/graphics/composer/2.1/utils/hwc2on1adapter/HWC2On1Adapter.cpp b/graphics/composer/2.1/utils/hwc2on1adapter/HWC2On1Adapter.cpp
index 6551a99a..5cc31cc5 100644
--- a/graphics/composer/2.1/utils/hwc2on1adapter/HWC2On1Adapter.cpp
+++ b/graphics/composer/2.1/utils/hwc2on1adapter/HWC2On1Adapter.cpp
@@ -2515,6 +2515,14 @@ Error HWC2On1Adapter::setAllDisplays() {
             continue;
         }
 
+    auto numLayers = mHwc1Contents[hwc1Id]->numHwLayers;
+ 
+        auto& clientTargetLayer = mHwc1Contents[hwc1Id]->hwLayers[numLayers - 1];
+        if (clientTargetLayer.compositionType == HWC_FRAMEBUFFER_TARGET && clientTargetLayer.acquireFenceFd != -1) {
+            close(clientTargetLayer.acquireFenceFd);
+            clientTargetLayer.acquireFenceFd = -1;
+        }
+
         auto displayId = mHwc1DisplayMap[hwc1Id];
         auto& display = mDisplays[displayId];
         auto retireFenceFd = mHwc1Contents[hwc1Id]->retireFenceFd;

